### DISCUSSION TITLE
Logisim-evolution internal architecture..

### DISCUSSION THREAD

**Discussion by efecollie** [logisim-evolution/logisim-evolution#1632] [Q&A] [answered]
Hi,
Is there any documentation on the internal structure/architecture/algorithms of logisim-evolution? I am planning to read its source code and I'd like to have some bird's eye view before delving deeper.. If not, can I get some advice on where to start reading?

**Comment by maehne** [comment, marked as answer]
There is no documentation of the internal structure/architecture of Logisim-evolution that I am aware of. However, there is the chapter "Value propagation" in the "Guide to Being a Logisim User", which is outlining Logisim's simulation algorithm and shortcomings. Help with improving all aspects of our documentation is highly welcome.

To understand a software project, I typically try to first get an overview of its project file hierarchy and how the software can be built. Then, I try to pick one small aspect of the software I would like to understand/improve at a time, find a starting point where I can interact with this aspect in the running software itself and then try to find implementation files, which seem to be associated to start from there my analysis of the underlying code.

If you can state a particular topic you would like to understand in more detail, we might be able to answer more precisely.

**Reply by newhinton** [reply]
I do have a specific question regarding the simulation!

Let's say i have two Circuits:
1. Complicated heavy duty circuit (CPU in my case)
and
2. Short Circuit that does side calculations

When i use two clocks, one that toggles on every tick, and one that only toggles every 200 ticks, will both circuits recalculate on every tick?
The goal to my question is to know if i can "turn off" my slow circuit by running its clock slower, so that i can quickly perform the task of the second circuit without it beeing slowed down by the first circuit.

**Reply by maehne** [reply]
Simulation is event-driven, i.e., only those component instances will get activated to update their outputs (i.e., generate new signal change events in case of different output values) if one of their inputs changed. However, currently each change triggers a redraw (see issue #786), which has a considerable performance penalty.
